File: practice/src/main/java/com/decade/practice/data/repositories/EdgeRepository.java
package com.decade.practice.data.repositories;

import com.decade.practice.model.domain.entity.Chat;
import com.decade.practice.model.domain.entity.Edge;
import com.decade.practice.model.domain.entity.User;
import com.decade.practice.utils.EventUtils;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.UUID;

public interface EdgeRepository extends JpaRepository<Edge, UUID> {

        List<Edge> findByOwner(User owner);

        @Query(
                "SELECT e FROM Edge e " +
                        "WHERE e.owner = :owner " +
                        "AND e.eventVersion <= :eventVersion " +
                        "AND e.head = TRUE"
        )
        List<Edge> getHeadEdge(
                @Param("owner") User owner,
                @Param("eventVersion") int eventVersion,
                Pageable pageable
        );

        @Query(
                "SELECT e FROM Edge e " +
                        "WHERE e.owner = :owner " +
                        "AND e.eventVersion <= :eventVersion " +
                        "AND e.dest=:to"
        )
        List<Edge> getEdgeTo(
                @Param("owner") User owner,
                @Param("eventVersion") int eventVersion,
                @Param("to") Chat to,
                Pageable pageable
        );

        @Query(
                "SELECT e FROM Edge e " +
                        "WHERE e.owner = :owner " +
                        "AND e.eventVersion <= :eventVersion " +
                        "AND e.from=:from"
        )
        List<Edge> getEdgeFrom(
                @Param("owner") User owner,
                @Param("eventVersion") int eventVersion,
                @Param("from") Chat from,
                Pageable pageable
        );

        // Convenience methods to replace Kotlin extension functions
        default Edge getHeadEdge(User owner, int version) {
                List<Edge> edges = getHeadEdge(owner, version, EventUtils.headEvent);
                return edges.isEmpty() ? null : edges.get(0);
        }

        default Edge getEdgeFrom(User owner, Chat from, int version) {
                List<Edge> edges = getEdgeFrom(owner, version, from, EventUtils.headEvent);
                return edges.isEmpty() ? null : edges.get(0);
        }

        default Edge getEdgeTo(User owner, Chat to, int version) {
                List<Edge> edges = getEdgeTo(owner, version, to, EventUtils.headEvent);
                return edges.isEmpty() ? null : edges.get(0);
        }
}

File: practice/src/main/java/com/decade/practice/database/repository/EdgeRepository.kt
package com.decade.practice.database.repository

import com.decade.practice.model.domain.entity.Chat
import com.decade.practice.model.domain.entity.Edge
import com.decade.practice.model.domain.entity.User
import com.decade.practice.utils.EventPageUtils
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import java.util.*

interface EdgeRepository : JpaRepository<Edge, UUID> {

      fun findByOwner(
            owner: User
      ): List<Edge>


      @Query(
            "SELECT e FROM Edge e " +
                    "WHERE e.owner = :owner " +
                    "AND e.eventVersion <= :eventVersion " +
                    "AND e.head = TRUE"
      )
      fun getHeadEdge(
            owner: User,
            eventVersion: Int,
            pageable: Pageable
      ): List<Edge>


      @Query(
            "SELECT e FROM Edge e " +
                    "WHERE e.owner = :owner " +
                    "AND e.eventVersion <= :eventVersion " +
                    "AND e.dest=:to"
      )
      fun getEdgeTo(
            owner: User,
            eventVersion: Int,
            to: Chat,
            pageable: Pageable
      ): List<Edge>

      @Query(
            "SELECT e FROM Edge e " +
                    "WHERE e.owner = :owner " +
                    "AND e.eventVersion <= :eventVersion " +
                    "AND e.from=:from"
      )
      fun getEdgeFrom(
            owner: User,
            eventVersion: Int,
            from: Chat,
            pageable: Pageable
      ): List<Edge>

}


fun EdgeRepository.getHeadEdge(
      owner: User,
      version: Int
): Edge? {
      return getHeadEdge(owner, version, EventPageUtils.headEvent).getOrNull(0)
}

fun EdgeRepository.getEdgeFrom(
      owner: User,
      from: Chat,
      version: Int
): Edge? {
      return getEdgeFrom(owner, version, from, EventPageUtils.headEvent).getOrNull(0)
}

fun EdgeRepository.getEdgeTo(
      owner: User,
      to: Chat,
      version: Int
): Edge? {
      return getEdgeTo(owner, version, to, EventPageUtils.headEvent).getOrNull(0)
}
